kotlinx-coroutines-core/common/src/channels/AbstractChannel.kt

```diff
@@ -832,7 +850,8 @@ internal abstract class AbstractChannel<E> : AbstractSendChannel<E>(), Channel<E
                     return@sc
                 }
                 if (result !== POLL_FAILED) {
-                    cont.resume(true)
+                    @Suppress("UNCHECKED_CAST")
+                    cont.resume(true, channel.onElementCancel?.bindCancellationFun(result as E, cont.context))
```

Comment:
Here is a general observation regarding `bindCancellationFun` and `onElementCancel`:

1) Both functions are invoked **rarely**, so the general performance cost doesn't matter here and allocation wrapper-lambda is ok
2) Most of the time `onElementCancel` is `null`.
3) Code with a lot of null-checks eats bytecode size here and there.
4) Maintaining code where `onElementCancel` always has to be wrapped into `bind` is not really pleasant


So my proposal is:
1) Create `bindCancellationFun` in the channel constructor. It will simplify our maintenance cost.
2) For regular channels, use singleton instance of `onElementCancel`. It will save our question marks and bytecode size.

After that, the code should become a bit more readable and smaller by the cost of additional allocation in a rare case of channel with resources

Reply:
`bindCancellationFun` is doing an actual _bind_ -- it captures the resulting value, so that the `onUnderlivedElement` function can be called for it in case of cancellation.